lwt/tls_lwt.ml

```diff
@@ -56,43 +75,47 @@ module Unix = struct
         (function
           | Out_of_memory -> raise Out_of_memory
           | exn -> (match t.state with
-              | `Error _ | `Eof -> ()
-              | `Active _ -> t.state <- `Error exn) ;
-            fail exn)
+              | `Error _ -> ()
+              | _ -> t.state <- `Error exn) ;
+            Lwt.reraise exn)
     in
     (recording_errors Lwt_cs.read, recording_errors Lwt_cs.write_full)
 
-  let when_some f = function None -> return_unit | Some x -> f x
+  let when_some f = function None -> Lwt.return_unit | Some x -> f x
 
   let rec read_react t =
 
     let handle tls buf =
       match Tls.Engine.handle_tls tls buf with
-      | Ok (state', `Response resp, `Data data) ->
-          let state' = match state' with
-            | `Ok tls  -> `Active tls
-            | `Eof     -> `Eof
-            | `Alert a -> `Error (Tls_alert a)
-          in
+      | Ok (state', eof, `Response resp, `Data data) ->
+          let state' = inject_state state' t.state in
+          let state' = Option.(value ~default:state' (map (fun `Eof -> half_close state' `read) eof)) in
           t.state <- state' ;
           safely (resp |> when_some (write_t t)) >|= fun () ->
           `Ok data
 
-      | Error (alert, `Response resp) ->
-          t.state <- `Error (Tls_failure alert) ;
+      | Error (fail, `Response resp) ->
+          t.state <- `Error (match fail with
+            | `Alert a -> Tls_alert a
+            | f -> Tls_failure f);
           write_t t resp >>= fun () -> read_react t
     in
 
     match t.state with
-    | `Error e  -> fail e
-    | `Eof      -> return `Eof
-    | `Active _ ->
-        read_t t t.recv_buf >>= fun n ->
-        match (t.state, n) with
-        | (`Active _  , 0) -> t.state <- `Eof ; return `Eof
-        | (`Active tls, n) -> handle tls (Cstruct.sub t.recv_buf 0 n)
-        | (`Error e, _)    -> fail e
-        | (`Eof, _)        -> return `Eof
+    | `Error e -> Lwt.reraise e
+    | `Closed
+    | `Read_closed _ -> Lwt.return `Eof
```

Comment:
If I look into `tls_mirage.ml`, the `` `Closed`` state produces a `` `Write `Closed`` where `tls_lwt.ml` produces a simple `` `Eof``. I'm in favor for the `tls_lwt.ml` version where a `ECONNRESET` can appear and should be handled as an `` `Eof``.

Reply:
It is tricky -- so in tls_lwt, a read when `` `Read_closed `` or `` `Closed `` is the internal state, and `` `Eof `` is signalled. If `writev` (or `write`) is called with `` `Closed `` as internal state, an exception is raised.

In mirage, I pushed 3ed1d0e to adhere to this behaviour. This is similar to how awa-ssh does it. :)

Thanks for your review.